wallet/reservation.c

```diff
@@ -0,0 +1,282 @@
+	json_add_utxos(cmd->ld->wallet, response, utxos);
+	json_array_end(response);
+
+	return command_success(cmd, response);
+}
+
+static const struct json_command unreserveinputs_command = {
+	"unreserveinputs",
+	"bitcoin",
+	json_unreserveinputs,
+	"Unreserve utxos",
+	false
+};
+AUTODATA(json_command, &unreserveinputs_command);
+
+static size_t utxos_weight(struct utxo **utxos)
+{
+	size_t weight;
+
+	weight = bitcoin_tx_core_weight(tal_count(utxos), 0);
+
+	for (size_t i = 0; i < tal_count(utxos); i++)
+		weight += utxo_spend_weight(utxos[i]);
+
+	return weight;
+}
+
+/* Returns how much more we need, or 0 if we have enough */
+static bool satisfied(u32 feerate_per_kw,
+		      struct amount_sat required,
+		      struct utxo **utxos,
+		      struct amount_sat *diff)
+{
+	struct amount_sat inamount;
+
+	/* If they ask for a huge amount, we'll never be satisfied anyway */
+	if (!amount_sat_add(&required, required,
+			    amount_tx_fee(feerate_per_kw, utxos_weight(utxos))))
+		return false;
+
+	inamount = AMOUNT_SAT(0);
+	for (size_t i = 0; i < tal_count(utxos); i++) {
+		/* If we ever allow them to specify random (unvetted) utxos,
+		 * this could happen.  We're satisfied in this case, I guess. */
+		if (!amount_sat_add(&inamount, inamount, utxos[i]->amount))
+			return true;
+	}
+
+	/* If inamount >= required, this succeeds. */
+	if (amount_sat_sub(diff, inamount, required))
+		return true;
+
+	if (!amount_sat_sub(diff, required, inamount))
+		abort();
+	return false;
+}
+
+static struct command_result *json_fundinputs(struct command *cmd,
+					      const char *buffer,
+					      const jsmntok_t *obj UNNEEDED,
+					      const jsmntok_t *params)
+{
+	struct json_stream *response;
+	struct utxo **utxos;
+	u32 *feerate_per_kw;
+	u32 *minconf;
+	struct amount_sat *amount, diff;
+
+	if (!param(cmd, buffer, params,
+		   p_req("satoshis", param_sat_or_all, &amount),
+		   p_req("feerate", param_feerate, &feerate_per_kw),
+		   p_opt_def("minconf", param_number, &minconf, 1),
+		   p_opt("utxos", param_utxos_any, &utxos),
+		   NULL))
+		return command_param_failed();
+
+	/* We start with their UTXOs, and keep adding until we meet their
+	 * output requirements. */
+	while (!satisfied(*feerate_per_kw, *amount, utxos, &diff)) {
+		struct utxo *utxo;
+
+		utxo = wallet_find_utxo(utxos, cmd->ld->wallet,
+					cmd->ld->topology->tip->height,
+					&diff,
+					*feerate_per_kw,
+					utxos);
+		if (utxo) {
+			tal_arr_expand(&utxos, utxo);
+			continue;
+		}
+
+		/* If they said "any", we expect to run out of utxos. */
+		if (amount_sat_eq(*amount, AMOUNT_SAT(-1ULL)))
+			break;
+
+		return command_fail(cmd, FUND_CANNOT_AFFORD,
+				    "Could not afford %s using all %zu available UTXOs: %s short",
+				    type_to_string(tmpctx,
+						   struct amount_sat,
+						   amount),
+				    tal_count(utxos),
+				    type_to_string(tmpctx,
+						   struct amount_sat,
+						   &diff));
+	}
+
+	response = json_stream_success(cmd);
```

Comment:
Is my understanding correct, that this command does not reserve the selected inputs?  Is it not better for it to atomically reserve the inputs as well, so that plugins do not have to do a `fundinputs`-`reserveinputs` loop?

Consider the case where two plugins need to get 0.1BTC each and the node owns two 0.2 BTC UTXOs, if we do not reserve atomically within `fundinputs`, then the plugins might end up getting the same UTXO from `fundinputs`, then reserving them with `reserveinputs`.  To be robust against this the plugins have to loop if `reserveinputs` fails, and only exit this loop if `fundinputs` fails or `reserveinputs` succeeds.

Reply:
Yes, a more detailed look at `wallet_find_utxo` shows that if multiple parallel `fundinputs` are invoked, they can indeed return the same UTXO sets in the resulting PSBTs.  Multiple plugins asking for similar amounts can end up having overlapping/same UTXO inputs.  So I strongly suggest to make `fundinputs` do reservation as well.  Otherwise plugins need to do a livelockable loop where they do `fundinputs`-`reserveinputs` in a loop until `reserveinputs` succeeds or `fundinputs` fails.